verifier/verifier.go

```diff
@@ -0,0 +1,494 @@
+				break
+			}
+		}
+	}
+
+	if invalidTimestamp {
+		return &notation.ValidationResult{
+			Error:  err,
+			Type:   trustpolicy.TypeAuthenticTimestamp,
+			Action: outcome.VerificationLevel.Enforcement[trustpolicy.TypeAuthenticTimestamp],
+		}
+	}
+
+	return &notation.ValidationResult{
+		Type:   trustpolicy.TypeAuthenticTimestamp,
+		Action: outcome.VerificationLevel.Enforcement[trustpolicy.TypeAuthenticTimestamp],
+	}
+}
+
+// verifyX509TrustedIdentities verified x509 trusted identities. This functions uses the VerificationResult from x509 trust store verification and modifies it
+func verifyX509TrustedIdentities(trustPolicy *trustpolicy.TrustPolicy, outcome *notation.VerificationOutcome, authenticityResult *notation.ValidationResult) {
+	// verify trusted identities
+	err := verifyX509TrustedIdentitiesCore(outcome.EnvelopeContent.SignerInfo.CertificateChain, trustPolicy)
+	if err != nil {
+		authenticityResult.Error = err
+	}
+}
+
+func executePlugin(ctx context.Context, installedPlugin plugin.Plugin, trustPolicy *trustpolicy.TrustPolicy, capabilitiesToVerify []proto.Capability, envelopeContent *signature.EnvelopeContent, pluginConfig map[string]string) (*proto.VerifySignatureResponse, error) {
+	// sanity check
+	if installedPlugin == nil {
+		return nil, errors.New("installedPlugin cannot be nil")
+	}
+
+	signerInfo, payloadInfo := &envelopeContent.SignerInfo, envelopeContent.Payload
+	var attributesToProcess []string
+	extendedAttributes := make(map[string]interface{})
+
+	for _, attr := range getNonPluginExtendedCriticalAttributes(signerInfo) {
+		extendedAttributes[attr.Key.(string)] = attr.Value
+		attributesToProcess = append(attributesToProcess, attr.Key.(string))
+	}
+
+	var certChain [][]byte
+	for _, cert := range signerInfo.CertificateChain {
+		certChain = append(certChain, cert.Raw)
+	}
+	var authenticSigningTime *time.Time
+	if signerInfo.SignedAttributes.SigningScheme == signature.SigningSchemeX509SigningAuthority {
+		authenticSigningTime = &signerInfo.SignedAttributes.SigningTime
+		// TODO use authenticSigningTime from signerInfo
+		// https://github.com/notaryproject/notation-core-go/issues/38
+	}
+
+	signature := proto.Signature{
+		CriticalAttributes: proto.CriticalAttributes{
+			ContentType:          payloadInfo.ContentType,
+			SigningScheme:        string(signerInfo.SignedAttributes.SigningScheme),
+			Expiry:               &signerInfo.SignedAttributes.Expiry,
+			AuthenticSigningTime: authenticSigningTime,
+			ExtendedAttributes:   extendedAttributes,
+		},
+		UnprocessedAttributes: attributesToProcess,
+		CertificateChain:      certChain,
+	}
+
+	policy := proto.TrustPolicy{
+		TrustedIdentities:     trustPolicy.TrustedIdentities,
+		SignatureVerification: capabilitiesToVerify,
+	}
+
+	req := &proto.VerifySignatureRequest{
+		ContractVersion: proto.ContractVersion,
```

Comment:
@JeyJeyGao Should `ContractVersion` always be set by the package `plugin` instead of the caller?

Reply:
Yes, it is. I'll update it in the next PR.